drools-game-engine-horserace-kjar/src/main/resources/rules/game/gamerules.drl

```diff
@@ -0,0 +1,55 @@
+package org.drools.game.horserace;
+
+import org.drools.game.core.api.*;
+import org.drools.game.model.api.*;
+import org.drools.game.horserace.model.*;
+import java.util.*;
+
+global GameMessageService messageService;
+global GameCallbackService callback;
+
+rule "Player Setup"
+    when
+        $player : Player()
+        not ( PlayerScore(player == $player) )
+    then
+        insert(  messageService.newGameMessage(  "Welcome, " + $player.getName() + ". Venture forth to the stables and mount a mighty steed!" ) );
+        insert(  new PlayerScore($player, 0) );
+end
+
+rule "general checkpoint"
+        lock-on-active true
+        when
+            $player : Player()
+            $playerscore : PlayerScore(player == $player)
+            $checkpoint : Checkpoint(players contains $player.getName())
+        then
+            insert(new CheckpointHit($player, $checkpoint.getOrder()));
+            insert(  messageService.newGameMessage(  "Contestant " + $player.getName() + " has crossed a checkpoint!" ) );
+end
+
+rule "passess general checkpoint hit"
+    lock-on-active true
+        when
+            $playerscore : PlayerScore()
+            $checkpoints : List(  ) from collect( Checkpoint() )
+            $checkpointhit : CheckpointHit(checkpointNumber == (($playerscore.currentCheckpoint + 1) % $checkpoints.size()))
```

Comment:
I use the modulus to wrap around. So, say the player passes through 0, 1, 2 in a race with three checkpoints. 2 + 1 % 3 = 0, which sets up the player to cross checkpoint 0 again.